handler/src/main/java/io/netty/handler/stream/ChunkedWriteHandler.java

```diff
@@ -167,21 +167,24 @@ private void discard(Throwable cause) {
             if (message instanceof ChunkedInput) {
                 ChunkedInput<?> in = (ChunkedInput<?>) message;
                 try {
-                    if (!in.isEndOfInput()) {
+                    // read state of the input in local variables before closing it
+                    boolean endOfInput = in.isEndOfInput();
+                    long inputLength = in.length();
+                    closeInput(in);
+                    if (!endOfInput) {
                         if (cause == null) {
                             cause = new ClosedChannelException();
                         }
                         currentWrite.fail(cause);
                     } else {
-                        currentWrite.success(in.length());
+                        currentWrite.success(inputLength);
                     }
-                    closeInput(in);
                 } catch (Exception e) {
+                    closeInput(in);
```

Comment:
@lutovich We assume here that only `in.isEndOfInput` or `in.length` were able to fail, right? I would probably move this try/catch upper to be more "targeted" and specific. Otherwise, it's not obvious from the first glance why do we try to close the input once again here.

Reply:
@kachayev yeah, I think this try-catch is to handle failures in `#isEndOfInput()` and `#length()`. Added a commit with a tighter try-catch as you suggest. It makes code a bit longer and adds a `continue`. Not sure it's an improvement but please take a look :)

Reply:
We need Python's try/catch/**else** :) We can mark that first block failed in a separate variable, and test its value instead of continue.